vernac/record.ml

```diff
@@ -322,46 +322,39 @@ let declare_projections indsp ctx ?(kind=Decls.StructureComponent) binder_name f
+                   else
+                     let ccl' = liftn 1 2 ccl in
+                     let p = mkLambda (x, lift 1 rp, ccl') in
+                     let branch = it_mkLambda_or_LetIn (mkRel nfi) lifted_fields in
+                     let ci = Inductiveops.make_case_info env indsp rci LetStyle in
+                     (* Record projections have no is *)
+                     mkCase (ci, p, mkRel 1, [|branch|]) in
+              let proj = it_mkLambda_or_LetIn (mkLambda (x,rp,body)) paramdecls in
+              let projtyp = it_mkProd_or_LetIn (mkProd (x,rp,ccl)) paramdecls in
+              try
+                let entry = Declare.definition_entry ~univs:ctx ~types:projtyp proj in
+                let kind = Decls.IsDefinition kind in
+                let kn = declare_constant ~name:fid ~kind (Declare.DefinitionEntry entry) in
                 Declare.definition_message fid;
-                kn, mkProj (Projection.make p false,mkRel 1)
-	      else
-		let ccl = subst_projection fid subst ti in
-		let body = match decl with
-                  | LocalDef (_,ci,_) -> subst_projection fid subst ci
-                  | LocalAssum ({binder_relevance=rci},_) ->
-	            (* [ccl] is defined in context [params;x:rp] *)
-		    (* [ccl'] is defined in context [params;x:rp;x:rp] *)
-		    let ccl' = liftn 1 2 ccl in
-                    let p = mkLambda (x, lift 1 rp, ccl') in
-		    let branch = it_mkLambda_or_LetIn (mkRel nfi) lifted_fields in
-                    let ci = Inductiveops.make_case_info env indsp rci LetStyle in
-                    (* Record projections have no is *)
-                    mkCase (ci, p, mkRel 1, [|branch|])
-                in
-		let proj =
-                  it_mkLambda_or_LetIn (mkLambda (x,rp,body)) paramdecls in
-		let projtyp =
-                  it_mkProd_or_LetIn (mkProd (x,rp,ccl)) paramdecls in
-	        try
-                  let entry = Declare.definition_entry ~univs:ctx ~types:projtyp proj in
-                  let kind = Decls.IsDefinition kind in
-                  let kn = declare_constant ~name:fid ~kind (Declare.DefinitionEntry entry) in
-		  let constr_fip =
-		    let proj_args = (*Rel 1 refers to "x"*) paramargs@[mkRel 1] in
-                      applist (mkConstU (kn,u),proj_args)
-                  in
-                  Declare.definition_message fid;
-		    kn, constr_fip
-                with Type_errors.TypeError (ctx,te) ->
-                  raise (NotDefinable (BadTypedProj (fid,ctx,te)))
+                if is_local_assum decl && primitive then
+                  let _ = DeclareInd.declare_primitive_projection p kn in
+                  let kn' = Projection.Repr.constant p in
+                  kn', mkProj (Projection.make p false,mkRel 1)
+                else
+                  let proj_args = (*Rel 1 refers to "x"*) paramargs@[mkRel 1] in
+                  kn, applist (mkConstU (kn,u),proj_args)
+              with Type_errors.TypeError (ctx,te) when not primitive ->
```

Comment:
Shouldn't we `try` only around the `declare_constant` rather than the whole block?
ie `let kn = try declare_constant ... with TypeError ... -> raise (NotDefinable ...) in`

Reply:
That's reasonable. I guess noone would object and I will do.

Reply:
Even better would be to remove the exception and use `result` , but that's orthogonal to this PR.